In [41]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [42]:
import os

# Carpeta de notebooks
ruta_proyecto = os.path.abspath("..")
print("Ruta proyecto:", ruta_proyecto)

# Listar todo lo que tiene
print("Contenido de proyecto_final:")
print(os.listdir(ruta_proyecto))


Ruta proyecto: c:\Users\ffmma\OneDrive\Desktop\proyecto_final
Contenido de proyecto_final:
['.git', '.gitignore', 'data', 'docs', 'notebooks', 'README.md', 'src']


In [43]:
import os

ruta_data = os.path.join("..", "data")
print("Contenido de data/:")
print(os.listdir(ruta_data))


Contenido de data/:
['data_processed', 'data_raw']


In [44]:
# Carpeta donde están los archivos
ruta_data_raw = os.path.join("..", "data", "data_raw")
ruta_data_raw

'..\\data\\data_raw'

In [45]:
archivo1 = os.path.join(ruta_data_raw, "telecom_clients_us.csv")

df1 = pd.read_csv(archivo1)

df1


user_id tariff_plan  date_start
0     166713           A  2019-08-15
1     166901           A  2019-08-23
2     168527           A  2019-10-29
3     167097           A  2019-09-01
4     168193           A  2019-10-16
..       ...         ...         ...
727   166554           B  2019-08-08
728   166911           B  2019-08-23
729   167012           B  2019-08-28
730   166867           B  2019-08-22
731   166565           B  2019-08-08

[732 rows x 3 columns]

In [46]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      732 non-null    int64 
 1   tariff_plan  732 non-null    object
 2   date_start   732 non-null    object
dtypes: int64(1), object(2)
memory usage: 17.3+ KB


In [47]:
#cambio de tipos de la columna date_start de objetct a datetime 
df1["date_start"]=pd.to_datetime(df1["date_start"], errors="coerce")

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   user_id      732 non-null    int64         
 1   tariff_plan  732 non-null    object        
 2   date_start   732 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 17.3+ KB


In [48]:
#chequo de valores nulos
df1.isnull().sum()
#ninguna columna tiene valores nulos 


user_id        0
tariff_plan    0
date_start     0
dtype: int64

In [49]:
df1.head()

user_id tariff_plan date_start
0   166713           A 2019-08-15
1   166901           A 2019-08-23
2   168527           A 2019-10-29
3   167097           A 2019-09-01
4   168193           A 2019-10-16

In [50]:
archivo2 = os.path.join(ruta_data_raw, "telecom_dataset_us.csv")

df2 = pd.read_csv(archivo2)

df2.head()

user_id                       date direction internal  operator_id  \
0   166377  2019-08-04 00:00:00+03:00        in    False          NaN   
1   166377  2019-08-05 00:00:00+03:00       out     True     880022.0   
2   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
3   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
4   166377  2019-08-05 00:00:00+03:00       out    False     880022.0   

   is_missed_call  calls_count  call_duration  total_call_duration  
0            True            2              0                    4  
1            True            3              0                    5  
2            True            1              0                    1  
3           False            1             10                   18  
4            True            3              0                   25

In [51]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              53902 non-null  int64  
 1   date                 53902 non-null  object 
 2   direction            53902 non-null  object 
 3   internal             53785 non-null  object 
 4   operator_id          45730 non-null  float64
 5   is_missed_call       53902 non-null  bool   
 6   calls_count          53902 non-null  int64  
 7   call_duration        53902 non-null  int64  
 8   total_call_duration  53902 non-null  int64  
dtypes: bool(1), float64(1), int64(4), object(3)
memory usage: 3.3+ MB


In [52]:
df2.sample(10)

user_id                       date direction internal  operator_id  \
28244   167180  2019-09-13 00:00:00+03:00       out     True     905930.0   
19788   166993  2019-10-02 00:00:00+03:00        in    False     901178.0   
31129   167402  2019-10-10 00:00:00+03:00        in    False          NaN   
31159   167402  2019-11-04 00:00:00+03:00        in    False     912010.0   
50689   168252  2019-11-21 00:00:00+03:00       out     True          NaN   
13983   166813  2019-09-19 00:00:00+03:00        in    False          NaN   
22260   167059  2019-10-23 00:00:00+03:00        in    False     911138.0   
6092    166536  2019-11-27 00:00:00+03:00       out    False     925828.0   
24857   167124  2019-10-17 00:00:00+03:00        in    False     901992.0   
48560   168187  2019-11-12 00:00:00+03:00       out    False     937902.0   

       is_missed_call  calls_count  call_duration  total_call_duration  
28244           False            2             11                   47  
19788           False            8            669                  720  
31129            True            8              0                  162  
31159           False            1             18                   56  
50689           False            1             53                   62  
13983            True            2              0                   61  
22260           False            6            631                  745  
6092            False            7            724                  819  
24857           False            1             36                   50  
48560            True           11              0                  413

In [53]:
#cambio a datetime de la columna operator_id
df2["date"]= pd.to_datetime(df2["date"],errors = "coerce")
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype                    
---  ------               --------------  -----                    
 0   user_id              53902 non-null  int64                    
 1   date                 53902 non-null  datetime64[ns, UTC+03:00]
 2   direction            53902 non-null  object                   
 3   internal             53785 non-null  object                   
 4   operator_id          45730 non-null  float64                  
 5   is_missed_call       53902 non-null  bool                     
 6   calls_count          53902 non-null  int64                    
 7   call_duration        53902 non-null  int64                    
 8   total_call_duration  53902 non-null  int64                    
dtypes: bool(1), datetime64[ns, UTC+03:00](1), float64(1), int64(4), object(2)
memory usage: 3.3+ MB


In [54]:
#extracción de fecha y horas de la columna date
df2["date"] = df2["date"].dt.date

df2


user_id        date direction internal  operator_id  is_missed_call  \
0       166377  2019-08-04        in    False          NaN            True   
1       166377  2019-08-05       out     True     880022.0            True   
2       166377  2019-08-05       out     True     880020.0            True   
3       166377  2019-08-05       out     True     880020.0           False   
4       166377  2019-08-05       out    False     880022.0            True   
...        ...         ...       ...      ...          ...             ...   
53897   168606  2019-11-10       out     True     957922.0            True   
53898   168606  2019-11-11       out     True     957922.0           False   
53899   168606  2019-11-15       out     True     957922.0           False   
53900   168606  2019-11-15       out     True     957922.0           False   
53901   168606  2019-11-19        in    False          NaN            True   

       calls_count  call_duration  total_call_duration  
0                2              0                    4  
1                3              0                    5  
2                1              0                    1  
3                1             10                   18  
4                3              0                   25  
...            ...            ...                  ...  
53897            1              0                   38  
53898            2            479                  501  
53899            4           3130                 3190  
53900            4           3130                 3190  
53901            2              0                   64  

[53902 rows x 9 columns]

In [55]:
#checar valores nulos
df2.isnull().sum()

user_id                   0
date                      0
direction                 0
internal                117
operator_id            8172
is_missed_call            0
calls_count               0
call_duration             0
total_call_duration       0
dtype: int64

En la columna internal se puede ver que solo hay 117 valores nulos, concidero que es una buena opción mantener estos ya que se puede también analizar si la falta de registro lleva algún patrón. 

La columna presenta un total de 8172 valores nulos, esto lo encuentro un poco preocupante ya que el principal objetivo es valorar el desempeño de los trabajadores. 
Considero que una buena opción para este caso cambiarlos por un valor "0", más adelante podrían ser huzados para poder mostrar la cantidad de llamadas con un operador sin id, para no perder datos y mantener un volumen aunque no se pueda identificar al operador. 

In [56]:
#convertí Nan a 0 
df2["operator_id"] = df2["operator_id"].fillna("Unknown")

df2.sample(5)

user_id        date direction internal operator_id  is_missed_call  \
1251    166405  2019-10-18       out    False    882686.0           False   
15679   166901  2019-09-16       out     True    896016.0            True   
28272   167180  2019-09-23       out    False    905932.0            True   
31050   167359  2019-11-27       out    False    920146.0           False   
24667   167112  2019-10-09       out    False    904014.0           False   

       calls_count  call_duration  total_call_duration  
1251            76           7764                 8891  
15679            4              0                    4  
28272            5              0                  186  
31050            3            461                  480  
24667           14            968                 1254

In [57]:
#hacer los float a int
df2["operator_id"] = df2["operator_id"].apply(
    lambda x: int(float(x)) if str(x).replace('.', '', 1).isdigit() else "Unknown"
)

df2.sample(10)


user_id        date direction internal operator_id  is_missed_call  \
18387   166974  2019-10-18       out    False      900354            True   
26317   167150  2019-10-23       out    False      905566            True   
47117   168185  2019-11-27       out    False      939212            True   
5487    166520  2019-10-25        in    False     Unknown            True   
11077   166680  2019-11-28       out    False      932502            True   
28072   167176  2019-11-13       out     True      905104           False   
9393    166658  2019-10-30       out    False      890404            True   
33627   167497  2019-10-21       out    False      921116            True   
35286   167532  2019-09-24       out     True      917850           False   
2392    166407  2019-10-09       out     True      907442            True   

       calls_count  call_duration  total_call_duration  
18387            4              0                  166  
26317            2              0                   62  
47117            2              0                   15  
5487            12              0                  248  
11077           31              0                  792  
28072            2            151                  156  
9393             1              0                    2  
33627            1              0                    9  
35286            1              0                   18  
2392             1              0                   22

In [58]:
df2.head()

user_id        date direction internal operator_id  is_missed_call  \
0   166377  2019-08-04        in    False     Unknown            True   
1   166377  2019-08-05       out     True      880022            True   
2   166377  2019-08-05       out     True      880020            True   
3   166377  2019-08-05       out     True      880020           False   
4   166377  2019-08-05       out    False      880022            True   

   calls_count  call_duration  total_call_duration  
0            2              0                    4  
1            3              0                    5  
2            1              0                    1  
3            1             10                   18  
4            3              0                   25

## Anlisis exploratorio (EDA)

In [59]:
#Resumen estadístico 
df1.describe(include="all")

user_id tariff_plan                     date_start
count      732.000000         732                            732
unique            NaN           3                            NaN
top               NaN           C                            NaN
freq              NaN         395                            NaN
mean    167431.927596         NaN  2019-09-14 16:39:20.655737600
min     166373.000000         NaN            2019-08-01 00:00:00
25%     166900.750000         NaN            2019-08-23 00:00:00
50%     167432.000000         NaN            2019-09-16 00:00:00
75%     167973.000000         NaN            2019-10-08 00:00:00
max     168606.000000         NaN            2019-10-31 00:00:00
std        633.810383         NaN                            NaN

Resumen estadístico general df1
#
Los identificadores de usuarios están completos, con todos los IDs caen dentro de un rángo que es lógico. La mayoría de los usuarios se encuentran en el plan C. Los ususarios se distribuyen de manera continua entre agosto y octubre de 2019. 

In [60]:
df2.describe(include="all")

user_id        date direction internal operator_id  \
count    53902.000000       53902     53902    53785       53902   
unique            NaN         119         2        2        1093   
top               NaN  2019-11-25       out    False     Unknown   
freq              NaN        1220     31917    47621        8172   
mean    167295.344477         NaN       NaN      NaN         NaN   
std        598.883775         NaN       NaN      NaN         NaN   
min     166377.000000         NaN       NaN      NaN         NaN   
25%     166782.000000         NaN       NaN      NaN         NaN   
50%     167162.000000         NaN       NaN      NaN         NaN   
75%     167819.000000         NaN       NaN      NaN         NaN   
max     168606.000000         NaN       NaN      NaN         NaN   

       is_missed_call   calls_count  call_duration  total_call_duration  
count           53902  53902.000000   53902.000000         53902.000000  
unique              2           NaN            NaN                  NaN  
top             False           NaN            NaN                  NaN  
freq            30334           NaN            NaN                  NaN  
mean              NaN     16.451245     866.684427          1157.133297  
std               NaN     62.917170    3731.791202          4403.468763  
min               NaN      1.000000       0.000000             0.000000  
25%               NaN      1.000000       0.000000            47.000000  
50%               NaN      4.000000      38.000000           210.000000  
75%               NaN     12.000000     572.000000           902.000000  
max               NaN   4817.000000  144395.000000        166155.000000

Podemos ver que la mayoría de las llamadas son salientes y externas, podría ser por que principalmente son llamadas de clientes externos. Una cantidad muy baja de llamadas no tiene operator_id y aproximadamente el 44% de las llamadas fueron perdidas. Vemos una variabilidad en la cantidad de llamadaspor regirtro algo grande, hay llamadas de 0 segundos y otras extremadamente largas lo que indica que la presencia de outliers podrían mostrar un sesgo. 


In [61]:
#Identificación de patrones entre variables para detección de operadores ineficientes 
resumen_operadores = df2.groupby("operator_id").agg(
    total_llamadas = ("calls_count", "sum"),
    llamadas_perdidas = ("is_missed_call", "sum"),
    llamadas_salientes = ("direction", lambda x: (x=="out").sum()),
    duracion_promedio = ("call_duration", "mean"),
    duracion_total = ("total_call_duration", "sum")
).reset_index()

resumen_operadores

operator_id  total_llamadas  llamadas_perdidas  llamadas_salientes  \
0         879896            1131                 50                 105   
1         879898            7974                100                 187   
2         880020              54                  7                  14   
3         880022             219                 33                  68   
4         880026            2439                 94                 179   
...          ...             ...                ...                 ...   
1088      972412              61                  2                   4   
1089      972460              70                  3                   7   
1090      973120               3                  1                   2   
1091      973286               2                  0                   0   
1092     Unknown          113716               8050                 200   

      duracion_promedio  duracion_total  
0            650.476562           97427  
1           1111.067729          391850  
2            104.090909            2624  
3            240.842105           22679  
4            856.939394          193666  
...                 ...             ...  
1088         699.600000            4548  
1089         257.888889            3122  
1090           2.500000              25  
1091          17.000000             105  
1092          66.013828         1910255  

[1093 rows x 6 columns]

In [62]:
#% de llamadas perdidas 
resumen_operadores["porcentaje_perdidas"] = (
    resumen_operadores["llamadas_perdidas"] / resumen_operadores["total_llamadas"] * 100
)

resumen_operadores.head(20)

operator_id  total_llamadas  llamadas_perdidas  llamadas_salientes  \
0       879896            1131                 50                 105   
1       879898            7974                100                 187   
2       880020              54                  7                  14   
3       880022             219                 33                  68   
4       880026            2439                 94                 179   
5       880028            2906                 91                 181   
6       880240              49                  8                  14   
7       881278              20                  7                   8   
8       882476               3                  0                   0   
9       882478               4                  1                   1   
10      882680             109                  3                   0   
11      882684            1364                 55                 113   
12      882686            3931                 88                 166   
13      882688              82                  3                   8   
14      882690            1766                 26                  45   
15      883018               2                  1                   1   
16      883898               1                  0                   1   
17      883940             694                 20                  36   
18      883942            2327                 67                 128   
19      884294               1                  0                   0   

    duracion_promedio  duracion_total  porcentaje_perdidas  
0          650.476562           97427             4.420866  
1         1111.067729          391850             1.254076  
2          104.090909            2624            12.962963  
3          240.842105           22679            15.068493  
4          856.939394          193666             3.854039  
5          746.265487          216250             3.131452  
6          275.928571            4436            16.326531  
7           58.125000             543            35.000000  
8           64.000000             287             0.000000  
9           22.666667             151            25.000000  
10         136.913043            8005             2.752294  
11        1422.175758          255782             4.032258  
12        1689.527426          469596             2.238616  
13         874.818182           11527             3.658537  
14        2058.942029          182846             1.472254  
15           0.000000              15            50.000000  
16          70.000000              80             0.000000  
17         510.139535           34637             2.881844  
18         914.994118          186818             2.879244  
19           5.000000              11             0.000000

In [63]:
resumen_operadores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1093 entries, 0 to 1092
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   operator_id          1093 non-null   object 
 1   total_llamadas       1093 non-null   int64  
 2   llamadas_perdidas    1093 non-null   int64  
 3   llamadas_salientes   1093 non-null   int64  
 4   duracion_promedio    1093 non-null   float64
 5   duracion_total       1093 non-null   int64  
 6   porcentaje_perdidas  1093 non-null   float64
dtypes: float64(2), int64(4), object(1)
memory usage: 59.9+ KB


Dado que el volumen de datos es muy grande, el análisis se vuelve más útil si nos enfocamos en comparar a los operadores con mayor porcentaje de llamadas perdidas frente a aquellos con menor porcentaje. Esto nos permite detectar diferencias claras de desempeño y entender qué prácticas, cargas de trabajo o factores podrían estar influyendo en la eficiencia de la atención.

In [68]:
operadores_100 = resumen_operadores[resumen_operadores['porcentaje_perdidas'] == 100]

operadores_100.head(21)

operator_id  total_llamadas  llamadas_perdidas  llamadas_salientes  \
27        885682               3                  3                   3   
160       899076               2                  2                   2   
253       905480               1                  1                   1   
381       914272               1                  1                   1   
386       914626               1                  1                   1   
528       924572               1                  1                   1   
569       928282               2                  2                   2   
617       932246               1                  1                   1   
687       937710               1                  1                   1   
688       937716               1                  1                   1   
697       937770               1                  1                   1   
698       937772               1                  1                   1   
700       937778               1                  1                   1   
715       937872               1                  1                   1   
809       941826               1                  1                   1   
970       954284               1                  1                   1   
1011      958458               1                  1                   1   
1064      969262               1                  1                   1   
1067      969284               1                  1                   1   
1077      970250               1                  1                   1   
1080      970258               1                  1                   1   

      duracion_promedio  duracion_total  porcentaje_perdidas  
27                  0.0              51                100.0  
160                 0.0              47                100.0  
253                 0.0               5                100.0  
381                 0.0               9                100.0  
386                 0.0               0                100.0  
528                 0.0               1                100.0  
569                 0.0              16                100.0  
617                 0.0              60                100.0  
687                 0.0              15                100.0  
688                 0.0              19                100.0  
697                 0.0              37                100.0  
698                 0.0              18                100.0  
700                 0.0              19                100.0  
715                 0.0              18                100.0  
809                 0.0               2                100.0  
970                 0.0               3                100.0  
1011                0.0               0                100.0  
1064                0.0              17                100.0  
1067                0.0               4                100.0  
1077                0.0               8                100.0  
1080                0.0              11                100.0

In [67]:
# Contar operadores únicos
ineficientes = operadores_100["operator_id"].nunique()
print("Operadores con un total de 100% de llamadas perdidas:")
print(ineficientes)


Operadores con un total de 100% de llamadas perdidas:
21


Al intentar hacer una comparación de % de llamadas perdidas entre los operadores podemos percatarnos que existen un total de 21 personal con un 100% de llamadas perdidas y muy pocas llamadas en el día. de 1 a 3 llamadas, deberá ponerse una atención especial a estos operadores ya que pueden ser un factor importante de riesgo en el desempeño.

In [70]:
#Filtro de mejores operadores para comparación 
mejores = resumen_operadores[(resumen_operadores["porcentaje_perdidas"] <= 5) & (resumen_operadores["total_llamadas"] >= 10)]
mejores.head(21)

operator_id  total_llamadas  llamadas_perdidas  llamadas_salientes  \
0       879896            1131                 50                 105   
1       879898            7974                100                 187   
4       880026            2439                 94                 179   
5       880028            2906                 91                 181   
10      882680             109                  3                   0   
11      882684            1364                 55                 113   
12      882686            3931                 88                 166   
13      882688              82                  3                   8   
14      882690            1766                 26                  45   
17      883940             694                 20                  36   
18      883942            2327                 67                 128   
24      884478            2361                 44                  79   
26      884946            2042                 42                  85   
28      885876           66049                135                 216   
29      885890           66016                110                 174   
31      886674              14                  0                   0   
36      887282            2452                 51                  97   
37      887416              22                  0                   0   
51      889754            1853                 72                 139   
54      890232              39                  0                  17   
56      890402            1928                 82                 163   

    duracion_promedio  duracion_total  porcentaje_perdidas  
0          650.476562           97427             4.420866  
1         1111.067729          391850             1.254076  
4          856.939394          193666             3.854039  
5          746.265487          216250             3.131452  
10         136.913043            8005             2.752294  
11        1422.175758          255782             4.032258  
12        1689.527426          469596             2.238616  
13         874.818182           11527             3.658537  
14        2058.942029          182846             1.472254  
17         510.139535           34637             2.881844  
18         914.994118          186818             2.879244  
24         549.096774           88310             1.863617  
26         460.913043           86776             2.056807  
28       14620.545455         6200335             0.204394  
29       13238.747292         5253549             0.166626  
31          16.818182             378             0.000000  
36        1456.720339          207280             2.079935  
37         163.571429            2659             0.000000  
51         548.846512          145726             3.885591  
54         986.352941           17297             0.000000  
56        1508.983193          388086             4.253112

Se tomó otra muestra de 21 operadores con mejor desempeño, similar a la muestra de los peores. En este grupo se observa que algunos operadores tienen un 0% de llamadas perdidas, aunque con un volumen relativamente bajo de llamadas (no menos de 14 en total), mientras que otros manejan un gran número de llamadas con un porcentaje muy bajo de pérdidas. Esto resalta un punto clave: es necesario revisar las prácticas y factores que permiten a estos operadores mantener un desempeño tan eficiente, ya que entenderlo puede ayudar a replicar estas buenas prácticas en otros operadores.